text/0000-road-to-typescript.md

```diff
@@ -0,0 +1,160 @@
+
+-->
+
+### Roadmap
+
+With these constraints in mind, the path to full TypeScript support in Ember requires RFCs to address each of the following points:
+
+- **The set of supported types.** Specifically, should the types support the whole API surface of Ember including both Classic and Octane features, or only Octane features? For example: mixins are not part of the Octane programming model *and* are a major source of the complexity of Ember’s current types; should they be included at all in types shipped with 
+
+- **Migration from DefinitelyTyped.** TypeScript users in the Ember community currently rely on the types in DefinitelyTyped. We need to define a transition story for them, which is closely related to the previous point. If we drop support for features currently supported on DefinitelyTyped, or make even well-motivated changes, how will users migrate successfully to them?
+
+- **Semantic Versioning of types and TypeScript compiler version support.** There is currently no widely-used definition for semantic versioning of TypeScript types; we must define the policy Ember uses—hopefully in a way which can benefit the broader TypeScript community. Additionally, we must define a support policy for TypeScript versions: do we pin to a minimum type when we release a major, do something our Node support policy, or some other approach? (See also further discussion [below](#semantic-versioning-and-supported-typescript-versions).)
+
+- **Supported compilation modes.** What should the out-of-the-box settings be for compiling Ember apps and addons written in TypeScript? This will need to be written with an eye to Embroider and the v2 Package Format as end goals, but without assuming them as available at the starting point. It must also address the transition path from current ember-cli-typescript (if any). It *must* address interoperation hazards between Babel and TypeScript: for example, in the decorators implementations, use of the `isolateModules` setting and associated lints, etc.
+
+- **The `@glimmer/component@2.0.0` Type API.** The current Glimmer component TypeScript API works well, but research efforts into “typed templates” and [component documentation](https://github.com/emberjs/rfcs/pull/678) have both exposed the importance of being able to specify more “type parameters” for components than simply its args, including its root element(s) and blocks, if any. Additionally, the types should be readily extensible if other such concerns emerge in the future.
+
+- **Official Documentation.** It is important, if Ember officially supports TypeScript, that we support it in all of Ember’s documentation, including guides, API docs, etc. This requires considerable design effort, including thinking through how to present them (side by side? toggle-able? etc.), where to introduce discussion of TypeScript support in the guides, etc.
```

Comment:
tabs for code snippets? With js and ts code?

Reply:
To be decided *in the docs RFC*. I'm intentionally leaving aside all of those details here. 😄